### ISSUE TITLE
Adaptation to the new openledger api

### ISSUE THREAD

**Issue by OpenLedgerApp** [bitshares/bitshares-ui#2572]
Openledger backend team implemented new api, which include these methods:
/exchanges (returns the list of supported exchange options, i.e what asset can we exchange to what. It contains information about gateway fee for each exchange option in the list.)
/exchanges/{exchange_id}/transfer/source/prototype?destination_address=<string>&destination_memo=<string> (returns to a user a "prototype" of a transfer that the user has to make in order to exchange his assets (that he sent) to another assets using given exchange option.)
/exchanges/{exchange_id}/transfer/destination (Method allows to check the validity of the transaction in the blockchain. This method must be called by browser/client when user wants to exchange one asset to another, and serves to check whether wanted operation in destination blockchain can be performed or not)

Need adaptation to this new api. But /exchanges returns incomplete list of exchanges, therefore some coins must work with old api.

**Comment by sschiessl-bcp** [comment]
The current implementation of the OpenLedger API still works?

The gateway code currently receives refactoring.

**Comment by MikhailKavalenka** [comment]
@sschiessl-bcp , yeah, the current implementation is still supported both on UI and BE.

**Comment by sschiessl-bcp** [comment]
Ok, this will likely take some time. Mid of next week we have a meeting to discuss our refactoring/streamlining of the gateway API. 

Do you have a high level document describing the exact definition/arguments/outputs of your new API?

**Comment by MikhailKavalenka** [comment]
[OL-gateways-api.pdf](https://github.com/bitshares/bitshares-ui/files/3068123/OL-gateways-api.pdf)
@sschiessl-bcp  here is the doc.

**Comment by startailcoon** [comment]
Pushed this to a milestone. Not added any hours to it as we need to look at the scope of this change requirements. 

- Do we need to refactor the gateway system to handle the new API by OL?
- Can we implement the new API in the current gateway system?

**Comment by dot5enko** [comment]
@MikhailKavalenka hi, sorry for such a long delay. 
I'm cool with what you're proposing, but have some thoughts on improvements of course:

1. merging assets and exchange methods into one uber request and call it `state` for example
2. adding `state_hash` which would be actually hash of given state or sequential number which increases with any changes like adding/removing exchange options/coins. 
3. add method which returns last `state_hash` only

So on the frontend we could keep `state_hash` and just compare it with what you have, making one small request (two if anything changes) per user's deposit/withdraw.

4. add method `limits/rates` which would return last rates, fees, and limits for `state`
.this method would always be called on withdraw/deposit

probably we should remove icons URLs from api return. as its really part of frontend. I don't actually 
understand why prototype and validation is not a one request. Probably we could merge them too. if operation is not permitted - just give an error. And yup - its better to return error codes instead of boolean flags on `"Validate Transfer" Method`

**Comment by dot5enko** [comment]
https://app.swaggerhub.com/apis/sdotsenko/BitsharesGatewayRequirements/2.0.0

this is the things i want to change i said above. Validate and prototype transfer methods looks good. I'll add them into docs later too 

Probably we could add optional `exchange_id` parameter to `limits` method. Cause probably its overweight to get all this info each time we want to withdraw or deposit something

@sschiessl-bcp, @startailcoon, @OpenLedgerApp  what do you think?

**Comment by sschiessl-bcp** [comment]
What is `active-wallets`? I can't see it in new specification

**Comment by dot5enko** [comment]
no `active-wallets` now. its solved with my `state_hash` solution. Openledger solution doesn't have this mechanism at all - they proposing as pulling all info every time i guess.

**Comment by sschiessl-bcp** [comment]
How do you incorporate a way for the gateway to deactivate a whole blockchain (former `active-wallet`, e.g. deactivate muse for maintenance)?

Will UI have to fetch all assets again?

**Comment by startailcoon** [comment]
I'm a little confused about the `state` and `state_hash`. Currently we check the trading pairs available (which are built from `/coins` and `/trading-pairs`) towards the exchanges `/active-wallet`. How does the new way handle a disabled asset deposit or withdrawal independent? 

Say that deposits from BTC -> open.BTC is disabled, but withdrawal from open.BTC -> BTC is still possible, what params would the gateway send in return?

**Comment by dot5enko** [comment]
guys, its possible that i've not covered everything. im open to discussion and improving my doc cause its a draft only. we are all here to make it comfortable for both us and gateways to use one api 


Just make some changes into https://app.swaggerhub.com/apis/sdotsenko/BitsharesGatewayRequirements/2.0.3. 

I've renamed endpoints a bit:
- the info which is rarely to change (assets and trading pairs, etc.) its a gateways info, so its a `info` endpoint. As its a huge request and its rarely changes it would be good to not load it each time. just comparing this info version: `info_hash` and only in case its not the same as client have - refresh gateway info ( this should happen when gateway added some trading pairs, or removed them. or fully disabled gateway )
- information that is frequently changed its a current gateway `state` - exchange pair fees, limits, disabled flags, etc. - endpoint `state`.  

@sschiessl-bcp we could add a list of active blockchains to `info` endpoint, or even into `state` as its not so big amount of info. Note that `Asset` model now have a `blockchain` property.  
@startailcoon look at `ExchangeOptions` model.  Its returned via `state` enpoint. BTC -> open.BTC and reverse direction are different exchanges and have their own ids. Thus one can be disabled and another enabled at same time.
